universal-application-tool-0.0.1/app/controllers/admin/AdminProgramController.java

```diff
@@ -0,0 +1,51 @@
+package controllers.admin;
+
+import javax.inject.Inject;
+import play.data.DynamicForm;
+import play.data.FormFactory;
+import play.mvc.Controller;
+import play.mvc.Http.Request;
+import play.mvc.Result;
+import services.program.ProgramService;
+import views.admin.ProgramList;
+import views.admin.ProgramNewOne;
+
+/**
+ * This controller contains an action to handle HTTP requests to the application's Admin "Create
+ * application" page.
+ */
+public class AdminProgramController extends Controller {
+
+  private final ProgramService service;
+  private final ProgramList listView;
+  private final ProgramNewOne newOneView;
+  private final FormFactory formFactory;
+
+  @Inject
+  public AdminProgramController(
+      ProgramService service,
+      ProgramList listView,
+      ProgramNewOne newOneView,
+      FormFactory formFactory) {
+    this.service = service;
+    this.listView = listView;
+    this.newOneView = newOneView;
+    this.formFactory = formFactory;
+  }
+
+  public Result list() {
+    return ok(listView.render(this.service.listProgramDefinitions()));
+  }
+
+  public Result newOne(Request request) {
+    return ok(newOneView.render(request));
+  }
+
+  public Result create(Request request) {
+    DynamicForm requestData = this.formFactory.form().bindFromRequest(request);
+    String name = requestData.get("name");
+    String description = requestData.get("description");
```

Comment:
Suggestion (blocking): recommend defining and using a [form class](https://www.playframework.com/documentation/2.8.x/JavaForms#Handling-form-submission) rather than `DynamicForm` when we know the structure of the form data this controller is meant to handle ahead of time.

Reply:
done!